File: backend/app/schemas/product.py
import uuid
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.catalog import BrandResponse, IngredientResponse, ProductTypeResponse, TaxonomyResponse
from app.schemas.review import ProductCanadaResponse, ProductImageResponse
from app.schemas.claims import ProductClaimCreate, ProductClaimResponse


ProductStatus = Literal[
    "draft",
    "imported",
    "processing",
    "review",
    "ready",
    "published",
    "active",
    "inactive",
    "discontinued",
]


class SkuCreate(BaseModel):
    sku: str = Field(min_length=1, max_length=100)
    barcode: str | None = Field(default=None, max_length=50)
    variant_name_en: str | None = None
    variant_name_zh: str | None = None
    net_quantity: float | None = Field(default=None, gt=0)
    quantity_unit: str | None = Field(default=None, max_length=20)


class SkuResponse(SkuCreate):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    product_id: uuid.UUID
    status: str


class ProductIngredientInput(BaseModel):
    ingredient_id: uuid.UUID
    position: int | None = Field(default=None, ge=1)
    is_key_ingredient: bool = False


class ProductIngredientResponse(BaseModel):
    ingredient: IngredientResponse
    position: int | None
    is_key_ingredient: bool


class ProductCreate(BaseModel):
    product_code: str = Field(min_length=1, max_length=50)
    brand_id: uuid.UUID
    product_type_id: uuid.UUID
    original_language: Literal["ko", "ja", "zh", "en", "fr", "other"]
    original_name: str = Field(min_length=1, max_length=500)
    product_name_en: str = Field(min_length=1, max_length=500)
    product_name_zh: str | None = None
    description_en: str | None = None
    description_zh: str | None = None
    how_to_use_en: str | None = None
    how_to_use_zh: str | None = None
    warnings_en: str | None = None
    warnings_zh: str | None = None
    country_of_origin: str | None = Field(default=None, min_length=2, max_length=2)
    source_inci: str | None = None
    attributes: dict = Field(default_factory=dict)
    skus: list[SkuCreate] = Field(default_factory=list)
    skin_type_ids: list[uuid.UUID] = Field(default_factory=list)
    skin_concern_ids: list[uuid.UUID] = Field(default_factory=list)
    ingredient_ids: list[uuid.UUID] = Field(default_factory=list)
    ingredients: list[ProductIngredientInput] | None = None
    claims: list[ProductClaimCreate] = Field(default_factory=list)


class ProductUpdate(BaseModel):
    brand_id: uuid.UUID | None = None
    product_type_id: uuid.UUID | None = None
    original_name: str | None = Field(default=None, min_length=1, max_length=500)
    product_name_en: str | None = Field(default=None, min_length=1, max_length=500)
    product_name_zh: str | None = None
    description_en: str | None = None
    description_zh: str | None = None
    how_to_use_en: str | None = None
    how_to_use_zh: str | None = None
    warnings_en: str | None = None
    warnings_zh: str | None = None
    source_inci: str | None = None
    attributes: dict | None = None
    country_of_origin: str | None = Field(default=None, min_length=2, max_length=2)
    skus: list[SkuCreate] | None = None
    skin_type_ids: list[uuid.UUID] | None = None
    skin_concern_ids: list[uuid.UUID] | None = None
    ingredient_ids: list[uuid.UUID] | None = None
    ingredients: list[ProductIngredientInput] | None = None
    claims: list[ProductClaimCreate] | None = None


class ProductStatusChange(BaseModel):
    status: ProductStatus


class ProductListItem(BaseModel):
    id: uuid.UUID
    product_code: str
    product_name_en: str
    brand_name: str
    product_type_name: str
    status: str
    compliance_status: str | None = None


class ProductResponse(BaseModel):
    id: uuid.UUID
    product_code: str
    original_language: str
    original_name: str
    product_name_en: str
    product_name_zh: str | None
    description_en: str | None
    description_zh: str | None
    how_to_use_en: str | None
    how_to_use_zh: str | None
    warnings_en: str | None
    warnings_zh: str | None
    country_of_origin: str | None
    source_inci: str | None
    attributes: dict
    status: str
    brand: BrandResponse
    product_type: ProductTypeResponse
    skus: list[SkuResponse]
    skin_types: list[TaxonomyResponse]
    skin_concerns: list[TaxonomyResponse]
    ingredients: list[ProductIngredientResponse]
    images: list[ProductImageResponse]
    canada: ProductCanadaResponse | None
    claims: list[ProductClaimResponse]

File: frontend/src/types/product.ts
export type ProductListItem = {
  id: string;
  product_code: string;
  product_name_en: string;
  brand_name: string;
  product_type_name: string;
  status: string;
  compliance_status: "pending" | "reviewing" | "approved" | "blocked" | null;
};

export type ProductListResponse = {
  items: ProductListItem[];
  total: number;
  offset: number;
  limit: number;
};

export type PriceSuggestionStatus = "pending" | "reviewing" | "approved" | "blocked";

import type { Brand, Ingredient, ProductType, Taxonomy } from "./catalog";

export type Sku = {
  id: string;
  product_id: string;
  sku: string;
  barcode: string | null;
  variant_name_en: string | null;
  variant_name_zh: string | null;
  net_quantity: number | null;
  quantity_unit: string | null;
  suggested_price: number | null;
  suggested_price_currency: string;
  suggested_price_status: PriceSuggestionStatus;
  suggested_price_note: string | null;
  suggested_price_reviewed_at: string | null;
  suggested_price_reviewed_by: string | null;
  status: string;
};

export type ProductImage = {
  id: string;
  product_id: string;
  image_url: string;
  image_type: string;
  sort_order: number;
  alt_text_en: string | null;
  alt_text_zh: string | null;
  status: string;
};

export type ProductCanada = {
  id: string;
  product_id: string;
  importer_name: string | null;
  distributor_name: string | null;
  canadian_label_status: string;
  cosmetic_notification_status: string;
  compliance_status: "pending" | "reviewing" | "approved" | "blocked";
  notes: string | null;
  reviewed_at: string | null;
};

export type Product = {
  id: string;
  product_code: string;
  original_language: string;
  original_name: string;
  product_name_en: string;
  product_name_zh: string | null;
  description_en: string | null;
  description_zh: string | null;
  how_to_use_en: string | null;
  how_to_use_zh: string | null;
  warnings_en: string | null;
  warnings_zh: string | null;
  country_of_origin: string | null;
  source_inci: string | null;
  attributes: Record<string, unknown>;
  status: string;
  brand: Brand;
  product_type: ProductType;
  skus: Sku[];
  skin_types: Taxonomy[];
  skin_concerns: Taxonomy[];
  ingredients: ProductIngredient[];
  images: ProductImage[];
  canada: ProductCanada | null;
  claims: ProductClaim[];
};

export type ProductIngredient = {
  ingredient: Ingredient;
  position: number | null;
  is_key_ingredient: boolean;
};

export type ProductClaim = {
  id: string;
  product_id: string;
  claim_type: string;
  claim_text_en: string;
  claim_text_zh: string | null;
  sort_order: number;
  status: string;
};

export type ReadinessCheck = { key: string; label: string; passed: boolean; weight: number; detail: string | null };
export type ProductReadiness = { product_id: string; score: number; ready: boolean; checks: ReadinessCheck[]; blockers: string[] };

export type ProductCreatePayload = {
  product_code: string;
  brand_id: string;
  product_type_id: string;
  original_language: string;
  original_name: string;
  product_name_en: string;
  product_name_zh: string;
  description_en: string;
  description_zh: string;
  how_to_use_en: string;
  how_to_use_zh: string;
  warnings_en: string;
  warnings_zh: string;
  country_of_origin: string;
  source_inci: string;
  attributes: Record<string, unknown>;
  skus: Array<{
    sku: string;
    barcode?: string;
    variant_name_en?: string;
    variant_name_zh?: string;
    net_quantity?: number;
    quantity_unit?: string;
    suggested_price?: number;
    suggested_price_currency?: string;
    suggested_price_note?: string;
  }>;
  skin_type_ids: string[];
  skin_concern_ids: string[];
  ingredient_ids: string[];
  ingredients?: Array<{ ingredient_id: string; position?: number; is_key_ingredient?: boolean }>;
  claims: Array<{ claim_type?: string; claim_text_en: string; claim_text_zh?: string; sort_order?: number }>;
};

export type ProductUpdatePayload = Partial<Pick<
  ProductCreatePayload,
  "product_name_en" | "product_name_zh" | "description_en" | "description_zh" | "how_to_use_en" | "how_to_use_zh" | "warnings_en" | "warnings_zh" | "source_inci" | "attributes" | "brand_id" | "product_type_id" | "original_name" | "country_of_origin" | "skus" | "skin_type_ids" | "skin_concern_ids" | "ingredient_ids" | "ingredients" | "claims"
>>;
